sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -103,17 +102,31 @@ case class ScalaUDF(
     }
   }
 
-  private def createToScalaConverter(i: Int, dataType: DataType): Any => Any = {
+  private def scalaConverter(i: Int, dataType: DataType): Any => Any = {
     if (inputEncoders.isEmpty) {
       // for untyped Scala UDF
-      CatalystTypeConverters.createToScalaConverter(dataType)
+      createToScalaConverter(dataType)
+    } else if (isPrimitive(dataType)) {
+      identity
     } else {
       val encoder = inputEncoders(i)
-      if (encoder.isDefined && encoder.get.isSerializedAsStructForTopLevel) {
-        val fromRow = encoder.get.resolveAndBind().createDeserializer()
-        row: Any => fromRow(row.asInstanceOf[InternalRow])
-      } else {
-        CatalystTypeConverters.createToScalaConverter(dataType)
+      encoder match {
+        case Some(enc) =>
+          if (enc.isSerializedAsStructForTopLevel) {
+            val fromRow = enc.resolveAndBind().createDeserializer()
+            row: Any => fromRow(row.asInstanceOf[InternalRow])
+          } else {
+            val child = children(i)
+            val attrs = new StructType().add(s"$child", child.dataType).toAttributes
```

Comment:
`child.toString` can be expensive. how about `"child"`? The name doesn't matter anyway.